C/mcal/SPI/SPI_program.c

```diff
@@ -0,0 +1,208 @@
+#include "../../common/Config.h"
+#include "../../common/Types.h"
+#include "../../common/Utils.h"
+#include "../../common/Registes.h"
+#include "SPI_private.h"
+#include "SPI_config.h"
+#include "SPI_interface.h"
+
+void SPI_MASTER_Init()
+{
+    #if MCU_TYPE == _AVR
+    SET_BIT(SPCR, SPCR_MSTR);
+    //DATA ORDER
+    #if SPI_DATA_ORDER == MSB_FIRST
+        CLR_BIT(SPCR, SPCR_DORD);
+    #elif SPI_DATA_ORDER == LSB_FIRST
+        SET_BIT(SPCR, SPCR_DORD);
+    #endif
+    //CLOCK POLARITY
+     #if SPI_CLK_POLARITY== IDLE_LOW
+        CLR_BIT(SPCR, SPCR_CPOL);
+    #elif SPI_CLK_POLARITY == IDLE_HIGH
+        SET_BIT(SPCR, SPCR_CPOL);
+    #endif
+    //CLOCK PHASE
+     #if SPI_CLK_PHASE== SAMPLE_FIRST
+        CLR_BIT(SPCR, SPCR_CPHA);
+    #elif SPI_CLK_PHASE == SETUP_FIRST
+        SET_BIT(SPCR, SPCR_CPHA);
+    #endif
+    //CLK FOR MASTER 128
+    SET_BIT(SPCR, SPCR_SPR0);
+    SET_BIT(SPCR, SPCR_SPR1);
+    CLR_BIT(SPSR, SPSR_SPI2X);
+    //SPI ENABLE
+    SET_BIT(SPCR, SPCR_SPE);
+    //INTERRUBT
```

Comment:
This function is really hard to read and to debug, so if you think the initialization is totally different tin `AVR` that `PIC` so in this case I believe the following implementation will be better
```
void SPI_MASTER_Init()
{
    #if MCU_TYPE == _AVR
    AVR_SPI_MasterInit();
   #elif MCU_TYPE == _PIC
    PIC_SPI_MasterIni();
   #else 
    #error "UNKOWN MCU"
    #endif 
````
where `AVR_SPI_MasterInit()` and `PIC_SPI_MasterInit()` are static helper functions that perform the initialization steps for specific mcu